FIX: Correctly select new ActionTree item after delete (ISX-1680)

timkeo opened the pull request:
### Description

Fix for [ISX-1680](https://jira.unity3d.com/browse/ISX-1680) - Deleting a Binding will correctly select a new item in the tree

### Changes made

Reworks the the logic to select a new Action/Binding after a delete operation by moving the functionality from Command.cs into ActionTreeView.cs.

Searches the ActionTreeView data to find the tree node that's about to be deleted and identify the siblings/parent node in the tree to select next and dispatches a separate Select command.

### Notes

Unlike Actions, the Binding IDs do not correspond to the indexes within the ActionTreeView and there's no way to determine a given Binding's siblings solely from the Action state. We must search through the visual tree to identify the new item to select (after delete) which is best executed within ActionTreeView.

For consistency (and safety should Action IDs change), this approach is applied both to Actions and Bindings.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

timkeo committed:
FIX: Correctly select new ActionTree item after delete (ISX-1680)

* Reworks the the logic to select a new Action/Binding after a delete operation

* Searches the ActionTreeView data to find the correct bindingIndex

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:
```diff
@@ -233,9 +233,24 @@ internal void AddComposite(string actionName, string compositeType)
+            }
+
+            // Deleting an item sometimes causes the UI Panel to lose focus; make sure we keep it
+            m_ActionsTreeView.Focus();
```
Interesting, I guess this happens if a binding is deleted and no other item is selected directrly after -> then the TreeView looses focus because nothing is selected. I think it's okay to just manual retrieve the Focus here. 

If I am right with my guess another way to solve this would be combining the commands (deleting and selecting) to avoid a refresh of the UI in between where there is no selection. (in my view not urgently necessary)

timkeo replied:
I wanted to avoid combining Delete/Select since I think they should be individual actions, but that was before discovering the UI refreshed between them. Maybe instead there's a way to "suspend" UI updates when executing multiple commands?

I do agree it's a minor issue and something can revisit later.

ritamerkl replied:
I think you can call the select command inside the delete command, what would mean we still have a delete and select command.

timkeo replied:
I played around with this a bit more to try and "combine" the Select/Delete operations and avoid needing to call `Focus()` on the TreeView. My idea was to create a "CommandList" implementation, that executed multiple commands (in sequence) and only fire the `StateChanged` event once at the end. [This commit](https://github.com/Unity-Technologies/InputSystem/commit/f7e317d1f10cccbff13dbe9d9616e909127e1521) holds my quick, dirty prototype.

Unfortunately, it didn't work; deleting multiple Action/Bindings in a row still causes the TreeView to (somehow) lose focus. I also found this issue exists in `develop` without my changes, and so It appears to be a deeper issue (possibly a bug in UITK).

So I'm inclined to keep the `Focus()` call to work-around the problem.

ritamerkl replied:
Interesting! Thank you for checking, I guess we are good to keep the Focus() work-around for now.

timkeo replied:
I opened [ticket ISX-1819](https://jira.unity3d.com/browse/ISX-1819) to follow up later.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:
```diff
@@ -180,19 +171,10 @@ public static Command DeleteBinding(int actionMapIndex, int bindingIndex)
-                index = Math.Max(state.selectedBindingIndex - 1, 0);
-            return state.SelectBinding(index);
+                // ActionsTreeView will dispatch a separate command to select the previous Binding
+                return state.With(selectedBindingIndex: -1);
```
How about just returning the state without changing the binding index? It will be overwritten by the ActionsTreeView anyway.

timkeo replied:
I was thinking the same thing but unsure about possible side-effects, and so thanks for suggesting this.
Fixed with fbacf6f958456ecc949d89a22634f77574cfc607.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:
```diff
@@ -157,19 +157,10 @@ public static Command DeleteAction(int actionMapIndex, string actionName)
-                index = Math.Max(state.selectedActionIndex - 1, 0);
-            return state.SelectAction(index);
+                // ActionsTreeView will dispatch a separate command to select the previous Action
+                return state.With(selectedActionIndex: -1);
```
How about just returning the state without changing the binding index? It will be overwritten by the ActionsTreeView anyway.

timkeo replied:
Fixed with fbacf6f958456ecc949d89a22634f77574cfc607.

ritamerkl commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:
```diff
@@ -293,6 +308,87 @@ private void OnValidateCommand(ValidateCommandEvent evt)
+                return treeView[index].data.name;
+            }
+
+            return string.Empty;
```
what happens if the action was not found/the treeview is empty and the string is empty? I think in this case nothing should get selected?

timkeo replied:
If `actionName` isn't found then `FindIndex()` returns -1 which gets incremented to 0 in the `else` branch causing the first Action in the list to be selected. Typically I don't like taking this kind of dependency on an API and prefer to write more "defensive" code, but in this case `List<T>.FindIndex()` is a well established .NET API that won't change.

Still, I added a comment to clarify this reasoning with fbacf6f958456ecc949d89a22634f77574cfc607.

ritamerkl replied:
And if the action list is empty? (Edge case, but still possible)

timkeo replied:
The outer condition `if (treeView.Count > 1)` handles this case, i.e. if the Action list is empty this will return 0 and we jump to the default return.
But even if we didn't check, `FindIndex()` would still return -1 if the list was empty (didn't match actionName).

ritamerkl reviewed (commented):
Good change. Just some thoughts to consider

ekcoh commented on Packages/com.unity.inputsystem/CHANGELOG.md:
```diff
@@ -58,6 +58,7 @@ however, it has to be formatted properly to pass verification tests.
 - Fixed case ISX-1436 (UI TK Input Action Asset Editor - Error deleting Bindings with DeleteKey on Windows).
 - Fixed issue with UI Toolkit based Input Action Editor not restoring it's selected items after Domain Reload.
 - Fixed the [`GetHapticCapabilitiesCommand`](xref:UnityEngine.InputSystem.XR.Haptics.GetHapticCapabilitiesCommand) always failing to execute due to a mismatch in the size in bytes of the payload and the size expected by XR devices. Changed [`HapticCapabilities`](xref:UnityEngine.InputSystem.XR.Haptics.HapticCapabilities) to include all properties returned by the XR input subsystem. This makes Inp ...
+- Fixed case ISX-1680 (UITK Input Action Asset Editor - Deleting a binding selects a seemingly random binding)
```
We typically try to avoid referencing internal ticket IDs since they make little sense to users since not publicly available while the release notes and the repo is. Note that ISXB tickets are referenced via public issue tracker though. I would rather suggest explaining the fixed scenario rather than reusing the ticket title, e.g.
"Fixed an issue where deleting a binding in the Input Action Editor while some binding items where not expanded would result in an unexpected selection change that may expand unrelated bindings in the tree view."

timkeo replied:
Updated CHANGELOG entry with fbacf6f958456ecc949d89a22634f77574cfc607.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:
```diff
@@ -293,6 +308,87 @@ private void OnValidateCommand(ValidateCommandEvent evt)
+                    {
+                        int prevIndex = Math.Max(childIndex - 1, 0);
+                        var node = parentNode.children.ElementAt(prevIndex);
+                        retVal = node.data.bindingIndex;
```
Shouldn't we return `retVal `directly here? Or is the loop guaranteed to break indirectly by `if (!action.hasChildren)` above when this condition is true?

timkeo replied:
Yes I'm missing a `break` here; thanks for the catch!
Fixed with fbacf6f958456ecc949d89a22634f77574cfc607.

ekcoh reviewed (approved):
Looks good, added some minor comments regarding CHANGELOG and abort condition for loop. Happy to see the "algorithm/psuedo comment" in GetPreviousBindingIndexFromViewTree since it makes it easier to read the code by first understanding selected approach.

Pauliusd01 commented:
Shouldn't composite bindings also have this behaviour? (Currently regular bindings select the binding above once deleted, composite bindings select nothing) This is also present in Develop, so It's not a regression or anything, just pointing out the differences

https://github.com/Unity-Technologies/InputSystem/assets/54306142/7a806b2d-c067-4f16-affc-0907cb717e0d

Pauliusd01 reviewed (approved):
LGTM. Checked deletion with the shortcut/context menu on different actions and bindings. Did not notice anything unexpected besides the mentioned inconsistency between regular bindings and composite bindings.

timkeo committed:
Respond to PR feedback

- Fix CHANGELOG message to remove (internal) ticket reference
- Delete command returns unmodified state
- Add missing 'break' from ActionTree search loop

timkeo commented:
I opened [a bug ticket](https://jira.unity3d.com/browse/ISX-1814) for the issue with Composite Bindings @Pauliusd01 mentioned.
I discovered it in my own testing but forgot to open the bug.

timkeo committed:
Merge branch 'develop' into isx-1680-fix-delete-binding
